File: scripts/code_review.py
"""Code review and automated fixing system for the codebase."""

import asyncio
import logging
import os
import subprocess
import time
from pathlib import Path
from typing import Dict, List, Optional

import aiohttp
import backoff

# Configure logging
logging.basicConfig(
    level=logging.INFO, format="%(asctime)s - %(name)s - %(levelname)s - %(message)s"
)
logger = logging.getLogger(__name__)

# Supported file extensions and their languages
SUPPORTED_LANGUAGES = {
    ".js": "JavaScript",
    ".jsx": "JavaScript",
    ".ts": "TypeScript",
    ".tsx": "TypeScript",
    ".py": "Python",
}


class CodeReviewSystem:
    """Manages code review and automated fixing."""

    def __init__(self):
        """Initialize the code review system."""
        self.openai_api_key = os.getenv("OPENAI_API_KEY")
        if not self.openai_api_key:
            raise ValueError("OPENAI_API_KEY environment variable is required")

        self.headers = {
            "Authorization": f"Bearer {self.openai_api_key}",
            "Content-Type": "application/json",
        }
        self.session = None
        self.rate_limit_remaining = 50  # Initial rate limit assumption
        self.rate_limit_reset = 0

    async def __aenter__(self):
        """Setup async context."""
        self.session = aiohttp.ClientSession()
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        """Cleanup async context."""
        if self.session:
            await self.session.close()

    @staticmethod
    def detect_language(file_path: str) -> Optional[str]:
        """Detect the programming language based on file extension."""
        _, ext = os.path.splitext(file_path)
        return SUPPORTED_LANGUAGES.get(ext)

    @backoff.on_exception(
        backoff.expo, (aiohttp.ClientError, asyncio.TimeoutError), max_tries=5
    )
    async def fix_code(self, file_path: str, language: str) -> Optional[str]:
        """Review and fix code using OpenAI API with improved error handling."""
        if not self.session:
            self.session = aiohttp.ClientSession()

        try:
            with open(file_path, "r", encoding="utf-8") as file:
                code_content = file.read()

            # Check rate limits
            if self.rate_limit_remaining <= 0:
                wait_time = max(0, self.rate_limit_reset - int(time.time()))
                if wait_time > 0:
                    logger.info(f"Rate limit reached. Waiting {wait_time} seconds...")
                    await asyncio.sleep(wait_time)

            prompt = f"""
            You are a code fixing expert. Analyze this {language} code and provide the complete fixed version that:
            1. Fixes any syntax errors and bugs
            2. Addresses security vulnerabilities
            3. Improves performance issues
            4. Ensures proper integration
            5. Follows best practices and style guidelines

            Return ONLY the complete fixed code without any explanations:

            {code_content}
            """

            async with self.session.post(
                "https://api.openai.com/v1/chat/completions",
                headers=self.headers,
                json={
                    "model": "gpt-4",
                    "messages": [{"role": "user", "content": prompt}],
                    "temperature": 0.3,
                },
                timeout=60,
            ) as response:
                # Update rate limits from headers
                self.rate_limit_remaining = int(
                    response.headers.get("x-ratelimit-remaining", 50)
                )
                self.rate_limit_reset = int(
                    response.headers.get("x-ratelimit-reset", 0)
                )

                if response.status == 200:
                    data = await response.json()
                    fixed_code = data["choices"][0]["message"]["content"]
                    logger.info(f"Successfully fixed {file_path}")
                    return fixed_code
                if response.status == 429:  # Rate limit exceeded
                    retry_after = int(response.headers.get("retry-after", 60))
                    logger.warning(
                        f"Rate limit exceeded. Waiting {retry_after} seconds..."
                    )
                    await asyncio.sleep(retry_after)
                    return await self.fix_code(file_path, language)
                logger.error(f"API error: {response.status} - {await response.text()}")
                return None

        except Exception as e:
            logger.error(f"Error fixing {file_path}: {str(e)}")
            return None

    @staticmethod
    def apply_linters(file_path: str, language: str) -> bool:
        """Apply language-specific linters and formatters."""
        try:
            if language == "Python":
                subprocess.run(["black", file_path], check=True, capture_output=True)
                subprocess.run(["flake8", file_path], check=True, capture_output=True)
            elif language in ["JavaScript", "TypeScript"]:
                subprocess.run(
                    ["npx", "eslint", "--fix", file_path],
                    check=True,
                    capture_output=True,
                )
                subprocess.run(
                    ["npx", "prettier", "--write", file_path],
                    check=True,
                    capture_output=True,
                )
            return True
        except subprocess.CalledProcessError as e:
            logger.warning(f"Linting warnings for {file_path}: {e.output.decode()}")
            return True  # Continue even with linting warnings
        except Exception as e:
            logger.error(f"Unexpected error during linting {file_path}: {str(e)}")
            return False

    @staticmethod
    def save_fixed_code(file_path: str, fixed_code: str) -> bool:
        """Save the fixed code back to the file."""
        try:
            with open(file_path, "w", encoding="utf-8") as file:
                file.write(fixed_code)
            logger.info(f"Successfully saved fixes to {file_path}")
            return True
        except Exception as e:
            logger.error(f"Error saving fixes to {file_path}: {str(e)}")
            return False

    async def process_directory(self, directory: str) -> Dict[str, List[str]]:
        """Process all supported files in the directory recursively."""
        results = {
            "fixed": [],
            "failed": [],
            "skipped": [],
            "timeout": [],
            "in_progress": [],
        }

        # Priority paths for processing
        priority_paths = {
            "critical": [  # Most critical core functionality files
                "backend/core/middleware_integration.py",
                "backend/core/config_manager.py",
                "backend/middleware/request_handler.py",
                "backend/config/base_config.py",
                "server/index.ts",
                "server/routes.ts",
            ],
            "highest": [  # Core system files
                "backend/core/",
                "backend/middleware/",
                "backend/config/",
                "backend/routes/",
            ],
            "high": [  # Main application files
                "src/App.",
                "src/main.",
                "static/js/flashcard-system.js",
                "static/js/study-system.js",
            ],
            "medium": [  # Supporting functionality
                "backend/services/",
                "backend/utils/",
                "src/components/",
                "src/hooks/",
            ],
            "low": ["backend/", "server/", "src/", "static/js/"],  # Non-critical files
        }

        # Directories to exclude
        exclude_dirs = {
            ".git",
            ".pythonlibs",
            "node_modules",
            "venv",
            "__pycache__",
            "migrations",
            "dist",
            "build",
            "coverage",
            "tests",
            ".pytest_cache",
            "logs",
            "temp",
            "tmp",
            ".venv",
        }

        # Maximum file size to process (1MB)
        MAX_FILE_SIZE = 1024 * 1024

        # Files to exclude
        exclude_files = {
            # Configuration files
            "migrations.py",
            "alembic.ini",
            "setup.py",
            "conftest.py",
            "jest.config.ts",
            "babel.config.js",
            "tsconfig.json",
            "vite.config.ts",
            "postcss.config.js",
            "tailwind.config.ts",
            "package.json",
            "package-lock.json",
            # Empty or boilerplate files
            "__init__.py",
            "index.d.ts",
            # Test files
            "test_*.py",
            "*_test.py",
            "*.test.ts",
            "*.spec.ts",
            "*.test.tsx",
            "*.spec.tsx",
            # Generated files
            "*.min.js",
            "*.min.css",
            "*.map",
            # Temporary files
            "*.tmp",
            "*.temp",
            "*.bak",
            "*.swp",
        }

        # Collect all eligible files first
        files_to_process = []
        total_files = 0
        skipped_files = {"size": [], "excluded": [], "unsupported": []}

        for root, dirs, files in os.walk(directory):
            # Skip excluded directories
            dirs[:] = [d for d in dirs if d not in exclude_dirs]

            for file_name in files:
                total_files += 1
                file_path = os.path.join(root, file_name)

                # Check for excluded files
                if any(
                    file_name.endswith(exc) or file_name == exc for exc in exclude_files
                ):
                    skipped_files["excluded"].append(file_path)
                    continue

                # Skip files that are too large
                try:
                    if os.path.getsize(file_path) > MAX_FILE_SIZE:
                        logger.info(f"Skipping large file: {file_path}")
                        skipped_files["size"].append(file_path)
                        continue
                except OSError:
                    logger.warning(f"Could not check size of {file_path}")
                    continue

                language = self.detect_language(file_path)
                if language:
                    # Check priority level (0-4, where 0 is critical and 4 is lowest)
                    priority = 4  # Default lowest priority

                    # First check exact matches for critical files
                    if file_path in priority_paths["critical"]:
                        priority = 0
                    else:
                        # Then check patterns from highest to lowest
                        for p_level, (category, patterns) in enumerate(
                            [
                                ("highest", priority_paths["highest"]),
                                ("high", priority_paths["high"]),
                                ("medium", priority_paths["medium"]),
                                ("low", priority_paths["low"]),
                            ],
                            1,
                        ):  # Start from 1 since 0 is reserved for critical
                            if any(pattern in file_path for pattern in patterns):
                                priority = p_level
                                break

                    # Only append if priority is not lowest
                    if priority < 4:
                        files_to_process.append((file_path, language, priority))
                        logger.info(f"Queued {file_path} with priority {priority}")
                    else:
                        skipped_files["excluded"].append(file_path)
                        logger.debug(f"Skipped low priority file: {file_path}")
                else:
                    results["skipped"].append(file_path)

        # Sort files with priority paths first, keeping priority info
        files_to_process.sort(key=lambda x: (x[2], x[0]))

        # Process files one at a time with careful error handling
        total_files = len(files_to_process)
        async with self as review_system:
            for index, (file_path, language, priority) in enumerate(files_to_process):
                results["in_progress"].append(file_path)
                logger.info(
                    f"\nProcessing file {index + 1}/{total_files} ({(index + 1)/total_files*100:.1f}%)"
                )
                logger.info(f"Current file: {file_path} (Priority: {priority})")
                try:
                    # Add delay between API calls to avoid rate limits
                    await asyncio.sleep(1)

                    # Step 1: Fix code
                    fixed_code = await review_system.fix_code(file_path, language)
                    if not fixed_code:
                        results["failed"].append(file_path)
                        continue

                    # Step 2: Save fixes
                    if review_system.save_fixed_code(file_path, fixed_code):
                        # Step 3: Apply linters
                        if review_system.apply_linters(file_path, language):
                            results["fixed"].append(file_path)
                            logger.info(f"Successfully processed {file_path}")
                        else:
                            results["failed"].append(file_path)
                    else:
                        results["failed"].append(file_path)

                except Exception as e:
                    logger.error(f"Error processing {file_path}: {str(e)}")
                    results["failed"].append(file_path)

                try:
                    # Set timeout based on priority level
                    timeout = {
                        0: 20,  # Critical files get shortest timeout
                        1: 30,  # Highest priority
                        2: 45,  # High priority
                        3: 60,  # Medium priority
                    }.get(priority, 90)  # Default longer timeout for other files

                    # Use asyncio.wait_for for the entire file processing
                    try:
                        async with asyncio.timeout(timeout):
                            logger.info(
                                f"Starting review of {file_path} with {timeout}s timeout"
                            )
                            fixed_code = await review_system.fix_code(
                                file_path, language
                            )

                            if not fixed_code:
                                logger.warning(f"No fixes generated for {file_path}")
                                results["failed"].append(file_path)
                                continue

                            # Save and lint in smaller steps with individual error handling
                            save_success = review_system.save_fixed_code(
                                file_path, fixed_code
                            )
                            if not save_success:
                                logger.error(f"Failed to save fixes for {file_path}")
                                results["failed"].append(file_path)
                                continue

                            lint_success = review_system.apply_linters(
                                file_path, language
                            )
                            if not lint_success:
                                logger.warning(f"Linting failed for {file_path}")
                                # Still mark as fixed if only linting failed
                                results["fixed"].append(file_path)
                            else:
                                results["fixed"].append(file_path)
                                logger.info(f"Successfully processed {file_path}")

                    except asyncio.TimeoutError:
                        logger.error(f"Timeout ({timeout}s) exceeded for {file_path}")
                        results["timeout"].append(file_path)
                    except Exception as e:
                        logger.error(
                            f"Unexpected error processing {file_path}: {str(e)}"
                        )
                        results["failed"].append(file_path)

                    # Adaptive delay based on priority and file size
                    file_size = os.path.getsize(file_path)
                    base_delay = {
                        0: 1,  # Critical files
                        1: 2,  # Highest priority
                        2: 3,  # High priority
                        3: 4,  # Medium priority
                    }.get(priority, 5)  # Default longer delay for other files

                    size_factor = min(
                        file_size / (500 * 1024), 2
                    )  # Cap at 2x for files larger than 500KB
                    delay = base_delay * size_factor

                    await asyncio.sleep(delay)

                except Exception as e:
                    logger.error(f"Error processing {file_path}: {str(e)}")
                    results["failed"].append(file_path)
                    results["in_progress"].remove(file_path)
                    await asyncio.sleep(5)
                    continue

                results["in_progress"].remove(file_path)

                # Progress update after each file
                logger.info(f"\nProgress Update:")
                logger.info(f"Fixed: {len(results['fixed'])} files")
                logger.info(f"Failed: {len(results['failed'])} files")
                logger.info(f"Timeout: {len(results['timeout'])} files")
                logger.info(f"In Progress: {len(results['in_progress'])} files")
                logger.info(f"Remaining: {total_files - (index + 1)} files")

                # Add periodic progress update
                if (index + 1) % 5 == 0:  # Update every 5 files
                    await asyncio.sleep(3)
                    percent_complete = ((index + 1) / total_files) * 100
                    logger.info(
                        f"Progress: {percent_complete:.1f}% ({index + 1}/{total_files} files)"
                    )
                    logger.info(
                        f"Status: Fixed={len(results['fixed'])}, Failed={len(results['failed'])}, Timeout={len(results['timeout'])}"
                    )

        # Log comprehensive processing summary
        logger.info("\nCode Review Summary:")
        logger.info("=" * 50)
        logger.info("Processing Statistics:")
        logger.info(f"Total files scanned: {total_files}")
        logger.info(f"Successfully fixed: {len(results['fixed'])} files")
        logger.info(f"Failed to process: {len(results['failed'])} files")
        logger.info(f"Timed out: {len(results['timeout'])} files")

        logger.info("\nSkipped Files:")
        logger.info(f"Size limit exceeded: {len(skipped_files['size'])}")
        logger.info(f"Excluded patterns: {len(skipped_files['excluded'])}")
        logger.info(f"Unsupported types: {len(skipped_files['unsupported'])}")

        # Calculate success rate
        processed_files = (
            len(results["fixed"]) + len(results["failed"]) + len(results["timeout"])
        )
        if processed_files > 0:
            success_rate = (len(results["fixed"]) / processed_files) * 100
            logger.info(f"\nSuccess Rate: {success_rate:.1f}%")

        logger.info("=" * 50)

        if results["failed"]:
            logger.error("\nFailed files:")
            for file in results["failed"]:
                logger.error(f"- {file}")

        return results


async def main():
    """Main entry point for the code fixing system."""
    try:
        review_system = CodeReviewSystem()
        project_root = Path(__file__).parent.parent

        logger.info("Starting automated code review process...")
        results = await review_system.process_directory(str(project_root))

        logger.info("\nCode Review Summary:")
        logger.info(f"Fixed files: {len(results['fixed'])}")
        logger.info(f"Failed files: {len(results['failed'])}")
        logger.info(f"Skipped files: {len(results['skipped'])}")

        if results["failed"]:
            logger.error("\nFailed files:")
            for file in results["failed"]:
                logger.error(f"- {file}")

    except Exception as e:
        logger.error(f"Fatal error in code review process: {str(e)}")
        raise


if __name__ == "__main__":
    asyncio.run(main())

File: scripts/code_review.ts
import { exec } from 'child_process';
import { promisify } from 'util';
import path, { dirname } from 'path';
import fs from 'fs/promises';
import { fileURLToPath } from 'url';

const execAsync = promisify(exec);
const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

const CORE_PATHS = [
  path.resolve(__dirname, '..', 'src'),
  path.resolve(__dirname, '..', 'server'),
  path.resolve(__dirname, '..', 'client'),
  path.resolve(__dirname, '..', 'db'),
];

// Prioritize these paths for initial review
const HIGH_PRIORITY_PATHS = [
  path.resolve(__dirname, '..', 'src/components'),
  path.resolve(__dirname, '..', 'server/services'),
  path.resolve(__dirname, '..', 'server/routes.ts'),
  path.resolve(__dirname, '..', 'db/schema.ts'),
];

const IGNORE_PATTERNS = [
  'node_modules',
  'dist',
  'build',
  '*.config.js',
  '*.config.ts',
  'postcss.config.js',
  'tailwind.config.ts',
  'vite.config.ts',
  '__tests__',
  'coverage',
  '.next',
  'public',
];

async function getTypeScriptFiles(path: string): Promise<string[]> {
  const allFiles: string[] = [];
  
  try {
    const stats = await fs.stat(path);
    
    if (stats.isFile()) {
      // If it's a single file and matches our pattern, add it
      if (/\.(ts|tsx|js|jsx)$/.test(path)) {
        allFiles.push(path);
      }
    } else if (stats.isDirectory()) {
      const entries = await fs.readdir(path, { withFileTypes: true });
      
      for (const entry of entries) {
        const fullPath = path + '/' + entry.name;
        
        // Skip ignored patterns
        if (IGNORE_PATTERNS.some(pattern => 
          entry.name.includes(pattern) || fullPath.includes(pattern)
        )) {
          continue;
        }
        
        if (entry.isDirectory()) {
          const subFiles = await getTypeScriptFiles(fullPath);
          allFiles.push(...subFiles);
        } else if (entry.isFile() && /\.(ts|tsx|js|jsx)$/.test(entry.name)) {
          allFiles.push(fullPath);
        }
      }
    }
  } catch (error) {
    console.error(`Error processing path ${path}:`, error);
  }
  
  return allFiles;
}

async function formatFiles(files: string[]): Promise<void> {
  console.log('Running Prettier...');
  try {
    await execAsync(`npx prettier --write ${files.join(' ')}`);
    console.log('Prettier formatting complete');
  } catch (error) {
    console.error('Error running Prettier:', error);
    throw error;
  }
}

async function lintFiles(files: string[]): Promise<void> {
  console.log('Running ESLint...');
  try {
    await execAsync(`npx eslint --fix ${files.join(' ')}`);
    console.log('ESLint fixes complete');
  } catch (error) {
    console.error('Error running ESLint:', error);
    throw error;
  }
}

async function generateReport(files: string[]): Promise<void> {
  console.log('\nCode Quality Report');
  console.log('==================');
  
  try {
    // Run ESLint in report mode with custom configuration for stricter analysis
    const { stdout: lintOutput } = await execAsync(
      `npx eslint ${files.join(' ')} --format stylish --max-warnings 0`
    );
    
    // Parse and categorize ESLint issues
    const issues = {
      warnings: (lintOutput.match(/warning/g) || []).length,
      errors: (lintOutput.match(/error/g) || []).length,
      unused: (lintOutput.match(/no-unused-vars/g) || []).length,
      hooks: (lintOutput.match(/react-hooks\//g) || []).length,
      typescript: (lintOutput.match(/@typescript-eslint\//g) || []).length
    };
    
    console.log('\nIssues Summary:');
    console.log('──────────────────────────────');
    console.log(`📊 Total Issues: ${issues.warnings + issues.errors}`);
    console.log(`🚨 Critical Errors: ${issues.errors}`);
    console.log(`⚠️  Warnings: ${issues.warnings}`);
    console.log(`🧹 Cleanup Needed:`);
    console.log(`  • Unused Variables/Imports: ${issues.unused}`);
    console.log(`  • React Hooks Issues: ${issues.hooks}`);
    console.log(`  • TypeScript-specific: ${issues.typescript}`);
    console.log('──────────────────────────────');
    
    if (issues.warnings > 0 || issues.errors > 0) {
      console.log('\nDetailed ESLint Report:');
      console.log(lintOutput);
    }
    
    // Code complexity analysis
    console.log('\nCode Complexity Analysis:');
    const complexityResults: Record<string, {
      cyclomaticComplexity: number;
      dependencies: string[];
      loc: number;
      functions: number;
      classes: number;
    }> = {};
    
    for (const file of files) {
      if (file.endsWith('.ts') || file.endsWith('.tsx')) {
        const { stdout: fileContent } = await execAsync(`cat ${file}`);
        
        // Calculate cyclomatic complexity (enhanced)
        const complexity = (fileContent.match(
          /(if|while|for|switch|catch|\?|&&|\|\||=>)/g
        ) || []).length;
        
        // Count lines of code (excluding comments and blank lines)
        const loc = fileContent
          .split('\n')
          .filter(line => line.trim() && !line.trim().startsWith('//'))
          .length;
        
        // Count functions and classes
        const functions = (fileContent.match(
          /\b(function|=>)\b|\b(get|set|async)\s+\w+\s*\(/g
        ) || []).length;
        
        const classes = (fileContent.match(/\bclass\s+\w+/g) || []).length;
        
        // Find imports
        const imports = fileContent.match(/import .+ from ['"][@\w\/\-\.]+['"]/g) || [];
        
        complexityResults[file] = {
          cyclomaticComplexity: complexity,
          dependencies: imports.map(imp => imp.match(/from ['"](.+)['"]/)?.[1] || ''),
          loc,
          functions,
          classes,
        };
      }
    }
    
    // Report high complexity files
    const highComplexityThreshold = 15; // Increased threshold for more meaningful results
    const highComplexityFiles = Object.entries(complexityResults)
      .filter(([_, data]) => data.cyclomaticComplexity > highComplexityThreshold)
      .sort(([_, a], [__, b]) => b.cyclomaticComplexity - a.cyclomaticComplexity);
    
    if (highComplexityFiles.length > 0) {
      console.log('\nHigh Complexity Files (needs refactoring):');
      highComplexityFiles.forEach(([file, data]) => {
        console.log(`- ${file}:`);
        console.log(`  Complexity: ${data.cyclomaticComplexity}`);
        console.log(`  Lines of Code: ${data.loc}`);
        console.log(`  Functions: ${data.functions}`);
        console.log(`  Classes: ${data.classes}`);
      });
    }
    
    // Dependency analysis
    console.log('\nDependency Analysis:');
    const dependencyUsage = new Map<string, number>();
    Object.values(complexityResults).forEach(({ dependencies }) => {
      dependencies.forEach(dep => {
        dependencyUsage.set(dep, (dependencyUsage.get(dep) || 0) + 1);
      });
    });
    
    // Report most used dependencies
    const topDependencies = Array.from(dependencyUsage.entries())
      .sort(([_, a], [__, b]) => b - a)
      .slice(0, 10);
    
    if (topDependencies.length > 0) {
      console.log('\nMost Used Dependencies:');
      topDependencies.forEach(([dep, count]) => {
        console.log(`- ${dep}: ${count} usages`);
      });
    }
    
    // Find potential dead code (files with no imports and low complexity)
    const unusedFiles = Object.entries(complexityResults)
      .filter(([file, data]) => {
        const noImports = !Array.from(dependencyUsage.keys()).some(dep => dep.includes(file));
        const lowComplexity = data.cyclomaticComplexity < 5;
        return noImports && lowComplexity;
      })
      .map(([file, _]) => file);
    
    if (unusedFiles.length > 0) {
      console.log('\nPotentially Unused/Dead Code:');
      unusedFiles.forEach(file => console.log(`- ${file}`));
    }
    
    // Enhanced file statistics
    console.log('\nFile Statistics:');
    console.log(`Total Files Processed: ${files.length}`);
    
    const typeStats = files.reduce((acc: Record<string, number>, file: string) => {
      const ext = path.extname(file);
      acc[ext] = (acc[ext] || 0) + 1;
      return acc;
    }, {});
    
    Object.entries(typeStats).forEach(([ext, count]) => {
      console.log(`${ext} files: ${count}`);
    });
    
    // Summary metrics with enhanced insights
    const totalLOC = Object.values(complexityResults).reduce((sum, data) => sum + data.loc, 0);
    const avgComplexity = Object.values(complexityResults)
      .reduce((sum, data) => sum + data.cyclomaticComplexity, 0) / Object.keys(complexityResults).length;
    
    console.log('\nCode Health Summary:');
    console.log('──────────────────────────────');
    
    console.log('\n📊 Overall Metrics:');
    console.log(`• Total Lines of Code: ${totalLOC}`);
    console.log(`• Average Complexity: ${avgComplexity.toFixed(2)}`);
    
    console.log('\n🚨 Critical Issues:');
    console.log(`• High Complexity Files: ${highComplexityFiles.length}`);
    console.log(`• Lint Warnings: ${issues.warnings}`);
    console.log(`• Lint Errors: ${issues.errors}`);
    
    console.log('\n🛠️  Maintenance Issues:');
    console.log(`• Unused Variables/Imports: ${issues.unused}`);
    console.log(`• Potential Dead Code Files: ${unusedFiles.length}`);
    
    const componentRatio = files.filter(f => f.includes('/components/')).length / files.length;
    console.log('\n📁 Code Organization:');
    console.log(`• Component to File Ratio: ${(componentRatio * 100).toFixed(1)}%`);
    console.log(`• TypeScript Files: ${files.filter(f => f.endsWith('.ts')).length}`);
    console.log(`• React Components: ${files.filter(f => f.endsWith('.tsx')).length}`);
    console.log('──────────────────────────────');
    
    console.log('\nRecommendations:');
    if (unusedFiles.length > 0) {
      console.log('\n1. Dead Code Removal:');
      console.log('   Consider removing or refactoring these potentially unused files:');
      unusedFiles.slice(0, 5).forEach(file => console.log(`   - ${file}`));
      if (unusedFiles.length > 5) {
        console.log(`   ... and ${unusedFiles.length - 5} more files`);
      }
    }
    
    if (highComplexityFiles.length > 0) {
      console.log('\n2. Complexity Reduction:');
      console.log('   These files need immediate refactoring:');
      highComplexityFiles.slice(0, 5).forEach(([file, data]) => 
        console.log(`   - ${file} (Complexity: ${data.cyclomaticComplexity}, Functions: ${data.functions})`)
      );
      if (highComplexityFiles.length > 5) {
        console.log(`   ... and ${highComplexityFiles.length - 5} more files`);
      }
    }
    
    if (issues.warnings > 0 || issues.errors > 0) {
      console.log('\n3. Code Quality Improvements:');
      console.log('   - Fix all ESLint warnings and errors');
      console.log('   - Remove unused variables and imports');
      console.log('   - Consider adding more TypeScript type definitions');
    }
    
    console.log('\nNext Steps:');
    if (issues.warnings > 0 || issues.errors > 0) {
      console.log('\nImmediate Actions:');
      console.log('1. Clean up unused variables in high-priority components:');
      console.log('   - Remove unused imports in AITutorAvatar.tsx (Sparkles, useEffect)');
      console.log('   - Clean up unused state variables in LearningPathVisualizer.tsx');
      console.log('   - Address React Hook dependencies in StudyTimer.tsx');
    }

    console.log('\nShort-term Improvements:');
    console.log('1. Remove or refactor dead code files (27 files identified)');
    console.log('2. Consolidate UI components and remove unused ones');
    console.log('3. Add error handling for unused error variables');

    console.log('\nLong-term Recommendations:');
    console.log('1. Implement comprehensive test coverage for components');
    console.log('2. Establish coding standards for variable naming and usage');
    console.log('3. Regular code review and cleanup sprints');
    
  } catch (error) {
    console.error('Error generating report:', error);
  }
}

async function main() {
  try {
    console.log('Starting automated code review...');
    
    // First process high priority paths
    let highPriorityFiles: string[] = [];
    for (const priorityPath of HIGH_PRIORITY_PATHS) {
      if (await fs.access(priorityPath).then(() => true).catch(() => false)) {
        console.log(`Processing high priority path: ${priorityPath}`);
        const files = await getTypeScriptFiles(priorityPath);
        highPriorityFiles.push(...files);
      }
    }

    // Process high priority files in smaller batches for more careful analysis
    const HIGH_PRIORITY_BATCH_SIZE = 10;
    if (highPriorityFiles.length > 0) {
      console.log('\nProcessing high priority files...');
      for (let i = 0; i < highPriorityFiles.length; i += HIGH_PRIORITY_BATCH_SIZE) {
        const batch = highPriorityFiles.slice(i, i + HIGH_PRIORITY_BATCH_SIZE);
        const batchNumber = Math.floor(i / HIGH_PRIORITY_BATCH_SIZE) + 1;
        const totalBatches = Math.ceil(highPriorityFiles.length / HIGH_PRIORITY_BATCH_SIZE);
        const progress = Math.round((batchNumber / totalBatches) * 100);
        console.log(`\n📦 Processing High Priority Batch ${batchNumber}/${totalBatches} (${progress}% complete)`);
        console.log(`   Processing ${batch.length} files in this batch...`);
        await formatFiles(batch);
        await lintFiles(batch);
      }
    }

    // Then process remaining core paths
    let remainingFiles: string[] = [];
    for (const corePath of CORE_PATHS) {
      if (await fs.access(corePath).then(() => true).catch(() => false)) {
        console.log(`Processing core path: ${corePath}`);
        const files = await getTypeScriptFiles(corePath);
        // Filter out already processed high priority files
        const newFiles = files.filter(file => !highPriorityFiles.includes(file));
        remainingFiles.push(...newFiles);
      }
    }

    // Process remaining files in larger batches for efficiency
    const BATCH_SIZE = 20;
    if (remainingFiles.length > 0) {
      console.log('\nProcessing remaining files...');
      for (let i = 0; i < remainingFiles.length; i += BATCH_SIZE) {
        const batch = remainingFiles.slice(i, i + BATCH_SIZE);
        const batchNumber = Math.floor(i / BATCH_SIZE) + 1;
        const totalBatches = Math.ceil(remainingFiles.length / BATCH_SIZE);
        const progress = Math.round((batchNumber / totalBatches) * 100);
        console.log(`\n📦 Processing Batch ${batchNumber}/${totalBatches} (${progress}% complete)`);
        console.log(`   Processing ${batch.length} files in this batch...`);
        await formatFiles(batch);
        await lintFiles(batch);
      }
    }
    
    // Generate report for all files
    const allFiles = [...highPriorityFiles, ...remainingFiles];
    console.log('\nGenerating comprehensive code quality report...');
    await generateReport(allFiles);
    
  } catch (error) {
    console.error('Error in automated code review:', error);
    process.exit(1);
  }
}

// Run if called directly
if (import.meta.url === new URL(process.argv[1], 'file:').href) {
  main().catch(console.error);
}

export { formatFiles, lintFiles, generateReport };